### ISSUE TITLE
CosmosDB where clause containing DateTime.UtcNow could not be translated

### ISSUE THREAD

**Issue by JonPSmith** [dotnet/efcore#18589]
If I filter a `DateTime` property in CosmosDB then it fails if I put the `DateTime.UtcNow`  within the LINQ expression, but doesn't fail if I put in a local variable set to `DateTime.UtcNow`. i.e.

```C#
var now = DateTime.UtcNow;
// The code below works
var books1 = await context.Books.Where(x => x.PublishedOn < now).ToListAsync();
//The code below throws an InvalidOperationException exception: not translatable.
var books2 = await context.Books.Where(x => x.PublishedOn < DateTime.UtcNow).ToListAsync(); 
```

NOTE: SQL works with DateTime.UtcNow in a LINQ where expression and I expected CosmosDB to do the same. But its so minor because of the workaround - I thought you should know anyway.

### Steps to reproduce

Entity 
``` C#
public class BookListNoSql
{
    [Key]
    public Guid BookId { get; set; }
    public string Title { get; set; }
    public DateTime PublishedOn { get; set; }
    public int YearPublished { get; set; }
    public decimal OrgPrice { get; set; }
    public decimal ActualPrice { get; set; }
    public string PromotionalText { get; set; }
    public string AuthorsOrdered { get; set; }
    public int ReviewsCount { get; set; }
    public double? ReviewsAverageVotes { get; set; }
}
```
DbContext
```c#
public class NoSqlDbContext : DbContext
{
    public NoSqlDbContext(DbContextOptions<NoSqlDbContext> options)
        : base(options) { }

    public DbSet<BookListNoSql> Books { get; set; }

}
```
Unit Test
NOTE: I am using Azure Cosmos DB Emulator
```C#
private readonly DbContextOptions<NoSqlDbContext> _options;

public TestNoSqlBookRead()
{
    var config = AppSettings.GetConfiguration();
    var builder = new DbContextOptionsBuilder<NoSqlDbContext>()
        .UseCosmos(config["endpoint"], config["authKey"], GetType().Name);
    _options = builder.Options;

    using var context = new NoSqlDbContext(_options);
    context.Database.EnsureCreated();
    if (context.Books.Select(_ => 1).AsEnumerable().Count() < 5)
    {
        var books = NoSqlTestData.CreateDummyBooks(10,stepByYear:true);
        context.AddRange(books);
        context.SaveChanges();
    }
}

[Fact]
public async Task TestDateFilterFailsDateTimeUtcNow()
{
    //SETUP
    using (var context = new NoSqlDbContext(_options))
    {
        //ATTEMPT
        var now = DateTime.UtcNow;
        var books = await context.Books.Where(x => x.PublishedOn < now).ToListAsync();
        var ex = await Assert.ThrowsAsync<InvalidOperationException>(async () =>
            await context.Books.Where(x => x.PublishedOn < DateTime.UtcNow).ToListAsync());

        //VERIFY
        books.Any().ShouldBeTrue();
        ex.Message.ShouldContain("could not be translated.");
    }
}
```

Got Exceptions? Include both the message and the stack trace
```
System.InvalidOperationException : The LINQ expression 'Where<BookListNoSql>(
    source: DbSet<BookListNoSql>, 
    predicate: (b) => b.PublishedOn < DateTime.UtcNow)' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to either AsEnumerable(), AsAsyncEnumerable(), ToList(), or ToListAsync(). See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.
   at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.<VisitMethodCall>g__CheckTranslated|8_0(ShapedQueryExpression translated, <>c__DisplayClass8_0& )
   at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass12_0`1.<ExecuteAsync>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetAsyncEnumerator(CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.ConfiguredCancelableAsyncEnumerable`1.GetAsyncEnumerator()
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
   at Test.UnitTests.ServiceLayer.TestNoSqlBookRead.TestDateFilterFailsDateTimeUtcNow() in C:\Users\Jon\Source\Repos\EfCoreSqlAndCosmos\Test\UnitTests\ServiceLayer\TestNoSqlBookRead.cs:line 55
--- End of stack trace from previous location where exception was thrown ---

```

### Further technical details

EF Core version: 3.0.0
Database provider: CosmosDB 3.0.0
Target framework:  .NET Core 3.0
Operating system:
IDE: Visual Studio 2019 16.3.1

**Comment by smitpatel** [comment]
Duplicate of #16143

**Comment by JonPSmith** [comment]
Really minor point so please ignore, but I mention it because the exception/message suggests a bug in the EF Core code.  In EF Core 5 rc2 you now get a `ArgumentNullException` (was `InvalidOperationException` in EF Core 3.1) and message "Value cannot be null. (Parameter 'instance')" (was "could not be translated" in EF Core 3.1)

Here is the stack trace for for the same code 
```
System.ArgumentNullException
Value cannot be null. (Parameter 'instance')
   at Microsoft.EntityFrameworkCore.Utilities.Check.NotNull[T](T value, String parameterName)
   at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosMemberTranslatorProvider.Translate(SqlExpression instance, MemberInfo member, Type returnType, IDiagnosticsLogger`1 logger)
   at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosSqlTranslatingExpressionVisitor.VisitMember(MemberExpression memberExpression)
   at System.Linq.Expressions.MemberExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosSqlTranslatingExpressionVisitor.VisitBinary(BinaryExpression binaryExpression)
   at System.Linq.Expressions.BinaryExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosSqlTranslatingExpressionVisitor.TranslateInternal(Expression expression)
   at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosSqlTranslatingExpressionVisitor.Translate(Expression expression)
   at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosQueryableMethodTranslatingExpressionVisitor.TranslateExpression(Expression expression)
   at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosQueryableMethodTranslatingExpressionVisitor.TranslateLambdaExpression(ShapedQueryExpression shapedQueryExpression, LambdaExpression lambdaExpression)
   at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosQueryableMethodTranslatingExpressionVisitor.TranslateWhere(ShapedQueryExpression source, LambdaExpression predicate)
   at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosQueryableMethodTranslatingExpressionVisitor.Visit(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass12_0`1.<ExecuteAsync>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetAsyncEnumerator(CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.ConfiguredCancelableAsyncEnumerable`1.GetAsyncEnumerator()
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
   at Test.UnitTests.ServiceLayer.TestNoSqlBookRead.TestDateFilterFailsDateTimeUtcNow() in C:\Users\JonPSmith\Source\Repos\EfCoreSqlAndCosmos\Test\UnitTests\ServiceLayer\TestNoSqlBookRead.cs:line 55
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_1.<<InvokeTestMethodAsync>b__1>d.MoveNext() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestInvoker.cs:line 264
--- End of stack trace from previous location where exception was thrown ---
   at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\ExecutionTimer.cs:line 48
   at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in C:\Dev\xunit\xunit\src\xunit.core\Sdk\ExceptionAggregator.cs:line 90

```

### Further technical details

EF Core version: 5.0.0 rc2
Database provider: CosmosDB  5.0.0 rc2
Target framework: .NET Core 3.1
Operating system: Windows
IDE: Visual Studio 2019 16.7

**Comment by smitpatel** [comment]
Incorrect null check. I will update.

**Comment by flacodirt** [comment]
> 
> 
> Really minor point so please ignore, but I mention it because the exception/message suggests a bug in the EF Core code. In EF Core 5 rc2 you now get a `ArgumentNullException` (was `InvalidOperationException` in EF Core 3.1) and message "Value cannot be null. (Parameter 'instance')" (was "could not be translated" in EF Core 3.1)
> 
> Here is the stack trace for for the same code
> 
> ```
> System.ArgumentNullException
> Value cannot be null. (Parameter 'instance')
>    at Microsoft.EntityFrameworkCore.Utilities.Check.NotNull[T](T value, String parameterName)
>    at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosMemberTranslatorProvider.Translate(SqlExpression instance, MemberInfo member, Type returnType, IDiagnosticsLogger`1 logger)
>    at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosSqlTranslatingExpressionVisitor.VisitMember(MemberExpression memberExpression)
>    at System.Linq.Expressions.MemberExpression.Accept(ExpressionVisitor visitor)
>    at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
>    at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosSqlTranslatingExpressionVisitor.VisitBinary(BinaryExpression binaryExpression)
>    at System.Linq.Expressions.BinaryExpression.Accept(ExpressionVisitor visitor)
>    at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
>    at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosSqlTranslatingExpressionVisitor.TranslateInternal(Expression expression)
>    at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosSqlTranslatingExpressionVisitor.Translate(Expression expression)
>    at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosQueryableMethodTranslatingExpressionVisitor.TranslateExpression(Expression expression)
>    at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosQueryableMethodTranslatingExpressionVisitor.TranslateLambdaExpression(ShapedQueryExpression shapedQueryExpression, LambdaExpression lambdaExpression)
>    at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosQueryableMethodTranslatingExpressionVisitor.TranslateWhere(ShapedQueryExpression source, LambdaExpression predicate)
>    at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
>    at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
>    at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
>    at Microsoft.EntityFrameworkCore.Cosmos.Query.Internal.CosmosQueryableMethodTranslatingExpressionVisitor.Visit(Expression expression)
>    at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
>    at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
>    at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
>    at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass12_0`1.<ExecuteAsync>b__0()
>    at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
>    at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query, CancellationToken cancellationToken)
>    at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)
>    at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetAsyncEnumerator(CancellationToken cancellationToken)
>    at System.Runtime.CompilerServices.ConfiguredCancelableAsyncEnumerable`1.GetAsyncEnumerator()
>    at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
>    at Test.UnitTests.ServiceLayer.TestNoSqlBookRead.TestDateFilterFailsDateTimeUtcNow() in C:\Users\JonPSmith\Source\Repos\EfCoreSqlAndCosmos\Test\UnitTests\ServiceLayer\TestNoSqlBookRead.cs:line 55
>    at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass48_1.<<InvokeTestMethodAsync>b__1>d.MoveNext() in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\Runners\TestInvoker.cs:line 264
> --- End of stack trace from previous location where exception was thrown ---
>    at Xunit.Sdk.ExecutionTimer.AggregateAsync(Func`1 asyncAction) in C:\Dev\xunit\xunit\src\xunit.execution\Sdk\Frameworks\ExecutionTimer.cs:line 48
>    at Xunit.Sdk.ExceptionAggregator.RunAsync(Func`1 code) in C:\Dev\xunit\xunit\src\xunit.core\Sdk\ExceptionAggregator.cs:line 90
> ```
> 
> ### Further technical details
> 
> EF Core version: 5.0.0 rc2
> Database provider: CosmosDB 5.0.0 rc2
> Target framework: .NET Core 3.1
> Operating system: Windows
> IDE: Visual Studio 2019 16.7

Thank you for explaining this, it is still happening in 5.0.2:

This throws error: `System.ArgumentNullException: Value cannot be null. (Parameter 'instance')`

`
var res = await _repo.GetAsync<ChatMessage>(x => x.CreatedDate >= DateTime.Now.AddHours(-5));
`

This works:

`
var time = DateTime.Now.AddHours(-5);
var res = await _repo.GetAsync<ChatMessage>(x => x.CreatedDate >= time);
`

**Comment by smitpatel** [comment]
This has been fixed in 6.0
